FD2.0 SetCommonRuntimeArgs() support w/ Multicast

kmabeeTT opened the pull request:
#4476: FD2.0 SetCommonRuntimeArgs() support w/ Multicast

     - Initial FD2 changes for SetCommonRuntimeArgs()
     - Allow FD2 to multicast common runtime args via SetCommonRuntimeArgs() instead of unicast
     - common_rt_args_offset must be 16B aligned for packed-write to work, so update
       validate_runtime_args_size() and get_common_runtime_args_offset() for this
     - BugFix in dataflow_api.h use L1_ARG_BASE instead of TRISC_L1_ARG_BASE in get_common_arg_addr()
     - Passes bunch of unit tests, with varying number of common runtime
       args across cores and processors.

     - Note: Tests not included here, WIP, will come in followup commit shortly.
     
     
     
     Edit: Updated to exclude tests for now, will include in follow up commit.

kmabeeTT commented:
Tagging @aliuTT and @abhullar-tt for review of feature code here.  Skip looking at tests, I am still working on tests not included here and see some issues in random test when I add readback of common-runtime-args (under debug) so I may have some issue.

aliuTT commented on tests/tt_metal/tt_metal/unit_tests/basic/runtime_args.cpp:
```diff
@@ -128,7 +128,7 @@ bool verify_results(
         // Verify common RT Args (same for all cores) if they exist.
         if (common_rt_args.size() > 0) {
-            const auto common_args_addr = rt_args_base_addr + common_rt_args_offset; // Common args are placed after unique args per core.
+            const auto common_args_addr = rt_args_base_addr + align(common_rt_args_offset, 16); // Common args are placed after unique args per core. 16B aligned.
```
use NOC_L1_ALIGNMENT

aliuTT replied:
We want to remove L1_ALIGNMENT and PCIE_ALIGNMENT. They're duplicates of `NOC_` versions.

aliuTT commented on tt_metal/impl/program/program.cpp:
```diff
@@ -840,7 +841,22 @@ void Program::update_runtime_args_transfer_info(Device* device) {
 
```
Does `kernel->cores_with_runtime_args()` need to be updated to not include multicast args?

kmabeeTT replied:
No, that is just for legacy/unicast runtime args to specific cores... with new common args, they are simply multicast to all coreranges that kernel is configured with (`kernel->logical_coreranges()`)

kmabeeTT committed:
#4476: FD2.0 SetCommonRuntimeArgs() support w/ Multicast

 - Initial FD2 changes for SetCommonRuntimeArgs()
 - Allow FD2 to multicast common runtime args via SetCommonRuntimeArgs() instead of unicast
 - common_rt_args_offset must be 16B aligned for packed-write to work, so update
   validate_runtime_args_size() and get_common_runtime_args_offset() for this
 - BugFix in dataflow_api.h use L1_ARG_BASE instead of TRISC_L1_ARG_BASE in get_common_arg_addr()
 - Passes bunch of unit tests, with varying number of common runtime
   args across cores and processors.

 - Note: Tests not included here, WIP, will come in followup commit shortly.
